### PR TITLE
5472 dbsync pipeline context

### PR DESCRIPTION
**Pull request by konangc87**
|Related issue|
|---|
|#5472|

## Description
Added pileline factory class and pipeline class to manage the sync row async.

## Tests

- Compilation without warnings in every supported platform
  - [ ] Linux
  - [ ] Windows
  - [ ] MAC OS X
- [ ] Source installation
- [ ] Package installation
- [ ] Source upgrade
- [ ] Package upgrade
- [ ] Review logs syntax and correct language
- [ ] QA templates contemplate the added capabilities

- Memory tests for Linux
  - [ ] Scan-build report
  - [ ] Coverity
  - [ ] Valgrind (memcheck and descriptor leaks check)
  - [ ] Dr. Memory
  - [ ] AddressSanitizer
- Memory tests for Windows
  - [ ] Scan-build report
  - [ ] Coverity
  - [ ] Dr. Memory
- Memory tests for macOS
  - [ ] Scan-build report
  - [ ] Leaks
  - [ ] AddressSanitizer

- [ ] Retrocompatibility with older Wazuh versions
- [ ] Working on cluster environments
- [ ] Configuration on demand reports new parameters
- [ ] The data flow works as expected (agent-manager-api-app)
- [ ] Added unit tests (for new features)
- [ ] Stress test for affected components

### CODE REVIEW

**Comment by konangc87** [comment]
UT:
image
Coverage:
image
valgrind:
image
cppcheck (legacy issues, no new issue):
image

**Comment by konangc87** [comment]
valgrind on testtool:
image

File: src/dbsync/src/dbsyncPipelineFactory.cpp [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+        , m_handle{ handle }
+        , m_txnContext{ DBSyncImplementation::instance().createTransaction(handle, tables) }
+        , m_maxQueueSize{ maxQueueSize }
+        , m_callback{ callback }
+        {
+            if (!m_callback || !m_handle || !m_txnContext)
+            {
+                throw dbsync_error
+                {
+                    3, "PipelineFactory, Invalid parameters."
```
Comment (tinachoB) [inline src/dbsync/src/dbsyncPipelineFactory.cpp:40]: Please, add a new std::make_pair in the db_exception and call it here.

File: src/dbsync/src/dbsync.cpp [hunk 2]
```diff
@@ -89,27 +90,36 @@ DBSYNC_HANDLE dbsync_create(const HostType     host_type,
...
 {
+    PipelineFactory::instance().release();
     DBSyncImplementation::instance().release();
 }
 
 TXN_HANDLE dbsync_create_txn(const DBSYNC_HANDLE handle,
                              const cJSON*        tables,
-                             const int           /*thread_number*/,
-                             const int           /*max_queue_size*/,
+                             const int           thread_number,
```
Comment (tinachoB) [inline src/dbsync/src/dbsync.cpp:99]: shouldn't these types be unsigned int ?

File: src/dbsync/src/dbsyncPipelineFactory.cpp [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+        std::lock_guard<std::mutex> lock{ m_contextsMutex };
+        const auto it
+        {
+            m_contexts.find(handle)
+        };
+        if (it == m_contexts.end())
+        {
+            throw dbsync_error
+            {
+                2, "PipelineFactory, Invalid handle value."
```
Comment (tinachoB) [inline src/dbsync/src/dbsyncPipelineFactory.cpp:175]: you should change this by:
throw dbsync_error { INVALID_HANDLE }; 

Or create a new pair with ID and description and call it here.

File: src/dbsync/src/dbsyncPipelineFactory.cpp [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
```
Comment (tinachoB) [inline src/dbsync/src/dbsyncPipelineFactory.cpp:191]: ditto

File: src/dbsync/src/dbsyncPipelineFactory.cpp [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+    }
+    PipelineCtxHandle PipelineFactory::create(const DBSYNC_HANDLE handle,
+                                              const char* tables,
+                                              const unsigned int threadNumber,
+                                              const unsigned int maxQueueSize,
+                                              const ResultCallback callback)
+    {
+        std::shared_ptr<IPipeline> spContext
+        {
+            new Pipeline
```
Comment (Dwordcito) [inline src/dbsync/src/dbsyncPipelineFactory.cpp:154]: could use make_shared on this line ?

File: src/dbsync/src/dbsyncPipelineFactory.cpp [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+        }
+        const std::shared_ptr<DispatchCallbackNode> m_spDispatchNode;
+        const std::shared_ptr<SyncRowNode> m_spSyncNode;
+        const DBSYNC_HANDLE m_handle;
+        const TXN_HANDLE m_txnContext;
+        const unsigned int m_maxQueueSize;
+        const ResultCallback m_callback;
+    };
+//----------------------------------------------------------------------------------------
+    PipelineFactory& PipelineFactory::instance()
```
Comment (Dwordcito) [inline src/dbsync/src/dbsyncPipelineFactory.cpp:136]: use noexcept for this function

File: src/dbsync/src/dbsyncPipelineFactory.cpp [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+    {
+        static PipelineFactory s_instance;
+        return s_instance;
+    }
+    void PipelineFactory::release()
```
Comment (Dwordcito) [inline src/dbsync/src/dbsyncPipelineFactory.cpp:141]: use noexcept for this function

File: src/dbsync/src/dbsyncPipelineFactory.cpp [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+    {
+        std::lock_guard<std::mutex> lock{ m_contextsMutex };
+        m_contexts.clear();
+    }
+    PipelineCtxHandle PipelineFactory::create(const DBSYNC_HANDLE handle,
```
Comment (Dwordcito) [inline src/dbsync/src/dbsyncPipelineFactory.cpp:146]: use noexcept for this function

File: src/dbsync/src/dbsyncPipelineFactory.cpp [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+        };
+        const auto ret { spContext.get() };
+        std::lock_guard<std::mutex> lock{ m_contextsMutex };
+        m_contexts.emplace(ret, spContext);
+        return ret;
+    }
+    const std::shared_ptr<IPipeline>& PipelineFactory::pipeline(const PipelineCtxHandle handle)
+    {
+        std::lock_guard<std::mutex> lock{ m_contextsMutex };
+        const auto it
```
Comment (Dwordcito) [inline src/dbsync/src/dbsyncPipelineFactory.cpp:167]: change to "const auto& it"

File: src/dbsync/src/dbsyncPipelineFactory.cpp [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+            {
+                INVALID_HANDLE
+            };
+        }
+        return it->second;
+    }
+    void PipelineFactory::destroy(const PipelineCtxHandle handle)
+    {
+        std::lock_guard<std::mutex> lock{ m_contextsMutex };
+        const auto it
```
Comment (Dwordcito) [inline src/dbsync/src/dbsyncPipelineFactory.cpp:183]: change to const auto& it

File: src/dbsync/src/dbsyncPipelineFactory.cpp [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+
+#include <tuple>
+#include "db_exception.h"
+#include "dbsyncPipelineFactory.h"
+#include "dbsync_implementation.h"
+#include "pipelineNodesImp.h"
+
+namespace DbSync
+{
+    class Pipeline : public IPipeline
```
[the remark this answers is not in the document]
Comment (konangc87) [reply src/dbsync/src/dbsyncPipelineFactory.cpp:20]: will do.

File: src/dbsync/src/dbsyncPipelineFactory.cpp (hunk 6 above)
Comment (konangc87) [reply src/dbsync/src/dbsyncPipelineFactory.cpp:136]: will do.

File: src/dbsync/src/dbsyncPipelineFactory.cpp (hunk 7 above)
Comment (konangc87) [reply src/dbsync/src/dbsyncPipelineFactory.cpp:141]: will do.

File: src/dbsync/src/dbsyncPipelineFactory.cpp (hunk 8 above)
Comment (konangc87) [reply src/dbsync/src/dbsyncPipelineFactory.cpp:146]: create can throw if dbsyncImplementation::createTransaction throws a json exception.

File: src/dbsync/src/dbsyncPipelineFactory.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (konangc87) [reply src/dbsync/src/dbsyncPipelineFactory.cpp:154]: will do.

File: src/dbsync/src/dbsyncPipelineFactory.cpp (hunk 9 above)
Comment (konangc87) [reply src/dbsync/src/dbsyncPipelineFactory.cpp:167]: will do.

File: src/dbsync/src/dbsyncPipelineFactory.cpp (hunk 10 above)
Comment (konangc87) [reply src/dbsync/src/dbsyncPipelineFactory.cpp:183]: will do.

File: src/dbsync/src/dbsyncPipelineFactory.h [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+    using TxnContext = void*;
+    using PipelineCtxHandle = void*;
+    struct IPipeline
+    {
+        virtual ~IPipeline() = default;
+        virtual void syncRow(const nlohmann::json& syncJson) = 0;
+        virtual void getDeleted(const ResultCallback callback) = 0;
+    };
+
+    class PipelineFactory
```
[the remark this answers is not in the document]
Comment (konangc87) [reply src/dbsync/src/dbsyncPipelineFactory.h:31]: will do.

**Pull request merged**